src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,313 @@
+package dotty.tools.dotc
+package transform
+
+import core.Types._
+import core.Contexts._
+import core.Flags._
+import ast.Trees._
+import ast.tpd
+import core.Decorators._
+
+
+/** space logic for checking exhaustivity and unreachability of
+  * pattern matching.
+  */
+
+
+/** space definition */
+sealed trait Space
+case object Empty extends Space
+case class Typ(t: Type) extends Space
+case class Kon(t: Type, params: List[Space]) extends Space
+case class Or(spaces: List[Space]) extends Space
+
+/** abstract space logic */
+trait SpaceLogic {
+  /** is `tp1` a subtype of `tp2` */
+  def isSubType(tp1: Type, tp2: Type): Boolean
+
+  /** is `tp1` the same type as `tp2` */
+  def isEqualType(tp1: Type, tp2: Type): Boolean
+
+  /** is `tp` a case class */
+  def isCaseClass(tp: Type): Boolean
+
+  /** is the type `tp` decompsable? i.e. all values of the type can be covered
+    * by its decomposed types.
+    *
+    * abstract sealed class and OrType can be decomposed
+    */
+  def canDecompose(tp: Type): Boolean
+
+  /** return parameters types of the case class `tp` */
+  def getSignature(tp: Type): List[Type]
+
+  /** get components of decomposable types */
+  def getPartitions(tp: Type): List[Type]
+
+  /** simplify space using the laws, there's no nested union after simplify */
+  def simplify(space: Space): Space = space match {
+    case Kon(tp, spaces) =>
+      val sp = Kon(tp, spaces.map(simplify _))
+      if (sp.params.exists(_ == Empty)) Empty
+      else sp
+    case Or(spaces) =>
+      spaces.map(simplify _).flatMap {
+        case Or(ss) => ss
+        case s => Seq(s)
+      } filter (_ != Empty) match {
+        case ss @ x::y::xs => Or(ss)
+        case s::Nil => s
+        case Nil => Empty
+      }
+    case Typ(tp) =>
+      if (canDecompose(tp) && getPartitions(tp).isEmpty) Empty
+      else space
+    case _ => space
+  }
+
+  /** is a subspace of b, equivalent to `a - b == Empty`, but faster */
```

Comment:
Missing the word `a`: It should be ``Is `a` a subtype of `b` ? This is equivalent to ...`